# Exercise 4.7: Deriving New Variables

# 01. Importing libraries

In [4]:
# Import libraries

import pandas as pd
import numpy as np
import os

# 02. Importing data

In [6]:
# Importing the merged orders-products data

path = r'/Users/vijaykasturi/Library/CloudStorage/OneDrive-Personal/Data Analytics Course - CareerFoundry/Data Immersion/A4/Instacart Basket Analysis'

df_ords_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [7]:
# Create a subset

df = df_ords_prods[:1000000]

In [8]:
# Define Function

def price_label(row):
    
    if row['prices'] <= 5:
        return 'Low-range Product'
    elif (row['prices'] > 5) and row['prices'] <= 15:
        return 'Mid-range Product'
    elif row['prices'] > 5:
        return 'High-range Product'
    else: return np.nan

In [9]:
# Apply the function

df['price_range'] = df.apply(price_label, axis=1)

/var/folders/73/ncnyv2zs7tlgp3vnng5pbtv00000gn/T/ipykernel_87387/93659844.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [10]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range Product     673183
Low-range Product     314392
High-range Product     12425
Name: count, dtype: int64

In [11]:
df['prices'].max()

99999.0

In [12]:
# Using loc()

df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

/var/folders/73/ncnyv2zs7tlgp3vnng5pbtv00000gn/T/ipykernel_87387/3838300506.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [13]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     673183
Low-range product     314392
High-range product     12425
Name: count, dtype: int64

In [14]:
df_ords_prods.loc[df_ords_prods['prices'] > 15, 'price_range_loc'] = 'High-range product'
df_ords_prods.loc[(df_ords_prods['prices'] <= 15) & (df_ords_prods['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 
df_ords_prods.loc[df_ords_prods['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [15]:
df_ords_prods['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [16]:
# Using for-loops

df_ords_prods['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [17]:
result = []

for value in df_ords_prods["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [18]:
# Insert "values" into data frame

df_ords_prods['busiest_day'] = result

In [19]:
df_ords_prods['busiest_day'].value_counts(dropna=False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# 04. Tasks 4.7

In [21]:
# 4.7.2 Changing the labels

result = []

for value in df_ords_prods["orders_day_of_week"]:
  if value == 0 or value == 1:
    result.append("Busiest days")
  elif value == 4 or value == 3:
    result.append("Slowest days")
  else:
    result.append("Regular days")

In [22]:
# Insert new into new column in data frame

df_ords_prods['daily_traffic_label'] = result

In [23]:
df_ords_prods['daily_traffic_label'].value_counts(dropna=False)

daily_traffic_label
Regular days    12916111
Busiest days    11864412
Slowest days     7624336
Name: count, dtype: int64

# 4.7.3: The 2 busiest days are 0 & 1 with agg. 11,864,412 orders; the 2 slowest days are 3 & 4 with agg. 7,624,336 orders.

In [25]:
# 4.7.4 Identifying busiest hours of the day

df_ords_prods['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [107]:
# Defining labels for Busiest periods

df_ords_prods.loc[df_ords_prods['order_hour_of_day'].isin(range(0,6)), 'busiest_period_of_day'] = 'Fewest orders'
df_ords_prods.loc[df_ords_prods['order_hour_of_day'].isin(range(10,19)), 'busiest_period_of_day'] = 'Most orders'
df_ords_prods['busiest_period_of_day'].fillna('Average orders', inplace = True)

/var/folders/73/ncnyv2zs7tlgp3vnng5pbtv00000gn/T/ipykernel_87387/3772987008.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_ords_prods['busiest_period_of_day'].fillna('Average orders', inplace = True)


In [115]:
# 4.7.5 Frequency for the new column

df_ords_prods['busiest_period_of_day'].value_counts(dropna=False)

busiest_period_of_day
Most orders       22388024
Average orders     9420507
Fewest orders       596328
Name: count, dtype: int64

# 05. Exporting data

In [117]:
# Export df_ords_prods as pkl

df_ords_prods.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_4.7.pkl'))

# End of Exercise 4.7